hexl/thread-pool/thread-pool.hpp

```diff
@@ -0,0 +1,310 @@
+// Copyright (C) 2022 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+#include <vector>
+
+#include "hexl/logging/logging.hpp"
+#include "hexl/util/check.hpp"
+#include "thread-pool/thread-pool-util.hpp"
+
+namespace intel {
+namespace hexl {
+
+using std::chrono::duration_cast;
+using tp_task_t = std::function<void(size_t id, size_t threads)>;
+
+class ThreadPool {
+ private:
+  // Properties
+  size_t total_threads = 0;                  // Total number of threads
+  std::atomic_uint64_t next_thread{0};       // Points to next free thread
+  std::vector<ThreadInfo*> thread_handlers;  // Thread's info
+  std::mutex pool_mutex;                     // Control pool's edition
+  bool setup_done = false;                   // Is thread pool initialized
+  inline static thread_local bool child;     // True if on child thread
+
+  // Methods
+
+  // StartThreads: Spawn a given number of threads
+  void StartThreads(size_t new_threads) {
+    size_t current_threads = total_threads;
+
+    // Add N new threads
+    for (size_t i = 0; i < new_threads; ++i) {
+      // Create handler and add it to vector of threads
+      ThreadInfo* thread_handler = new ThreadInfo();
+      thread_handlers.emplace_back(thread_handler);
```

Comment:
```suggestion
      ThreadInfo* thread_handler = thread_handlers.emplace_back(new ThreadInfo());
```
Apparently C++17 returns a ref to the inserted element !